File: python/structures/classes_and_methods.py
# class
class A:
    # function definition
    def f (self): # every method (function that will be called with an object)
                  # requires self as first parameter
                  # self is a reference to the object you invoke the method on
        ...       # using 'self' is convention - can use any word

def g ():         # outside of a class, functions do not need
                  #     a 'self' parameter

# dynamically typed
x = A()
y = A()
z = A()

isinstance(z, A)      # True

x.f() # self is x
y.f() # self is y

# inheritance
class B(A):           # class B inherits from class A
    ...

issubclass(B, A)      # True

# every class automatically inherits from object
issubclass(A, object) # True

# globals
h = 0

def f (...) :
    global h         # access the 'h' variable
    ...

File: javascript/structures/classes_and_methods.js
// class
class A {
    // function definition
    f() { }       // methods  - functions that belong to a class -
                  //  do not need the 'function' keyword
}

function g () { } // functions outside of classes need the 'function'
                  //  keyword

// dynamically typed
x = new A()       // need the 'new' keyword when constructing an object
y = new A()
z = new A()

z instanceof A    // true - instanceof only works for objects

/*********************************************************/

class B {
  constructor (a, b) {  // class constructor
    this.a = a
    this.b = b
  }
}

let b = new B(2, 3)   // need 'new' keyword to create new instance
b.a === 2             // access instance variables
b["a"] === 2
b.b === 3
b["b"] === 3
